### ISSUE TITLE
Asymmetric (two model) learning of pairs

### ISSUE THREAD

**Issue by turian** [UKPLab/sentence-transformers#3024]
My understanding is that fine-tuning using sentence-transformers over pairs assumes that the pairs are symmetric.
For example, using [ContrastiveTensionLossInBatchNegatives](https://sbert.net/docs/package_reference/sentence_transformer/losses.html#contrastivetensionlossinbatchnegatives).

This means that `embed(anchor1) = embed(anchor2)` and that `d(anchor1, anchor2) = d(anchor2, anchor1)`.

However, there are many use-cases where the order of the sentences is important. For example: `anchor1` is a SUMMARY of `anchor2`.

It would be great, optionally, to allow fine-tuning to create TWO fine-tuned models. One for embedding `anchor1` and the second for embedding `anchor2`. This would make it easy to fine-tune on asymmetric tasks.

**Comment by tomaarsen** [comment]
Hello!

There are indeed symmetric losses in Sentence Transformers, but also a few asymmetric ones. For example [MultipleNegativesRankingLoss](https://sbert.net/docs/package_reference/sentence_transformer/losses.html#multiplenegativesrankingloss), which is arguably the most common loss. This loss essentially trains:
* Given the anchor text, which other text from all positive (and optionally negative) values in the batch is the one that corresponds with the anchor?

If you have asymmetric data like summaries, then you might be able to train:
* Given a summary, which of the texts from the batch is the full document?

or
* Given a document, which of the texts from the batch is its summary?

depending on what you choose as your first column. The resulting model becomes adept at the asymmetric task (e.g. information retrieval).

---

Some model authors use prefixes/prompts for each type of data, so that the model can distinguish e.g. a query of "what are pandas" and a document of "what are pandas". There's some docs on them [here](https://sbert.net/examples/applications/computing-embeddings/README.html#prompt-templates), but note that this only refers to inference. For training you have to manually prepend the prompts.

Sentence Transformers does not support dual model setups right now, and I don't think I'll be moving in that direction soon. 

- Tom Aarsen

**Comment by turian** [comment]
@tomaarsen Thank you. Besides [MultipleNegativesRankingLoss](https://sbert.net/docs/package_reference/sentence_transformer/losses.html#multiplenegativesrankingloss), what are the other asymmetric losses?

We don't really have negatives for our task, unless we do randomly sampling. But I guess the batch setup does this anyway?

**Comment by tomaarsen** [comment]
> But I guess the batch setup does this anyway?
Yes, `MultipleNegativesRankingLoss` is an example of an "in-batch negatives" loss - i.e. it uses other values in the batch as the negative.
You can also use [`mine_hard_negatives`](https://sbert.net/docs/package_reference/util.html#sentence_transformers.util.mine_hard_negatives) to get negatives from your `(anchor, positive)` tuples, but this is not necessary. The idea is that if your dataset is big enough (or if you provide a large corpus), then it'll find negatives that are similar-ish but not really similar to the anchor. These "difficult cases" often result in a stronger model.

A lot of the losses are asymmetric, such as all losses that list `(anchor, positive)` or `(anchor, positive, negative)` in the [Loss Overview](https://sbert.net/docs/sentence_transformer/loss_overview.html). Some losses that are symmetric are the losses under `(sentence_A, sentence_B) pairs` with `float similarity score` labels. I could try and list them all, but it'd take a little bit to go through all of them again

- Tom Aarsen
